### PR TITLE
CHE-516: Fix unstable debbuger behavior

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
**_2 Upvotes**_ @sleshchenko @skabashnyuk @vparfonov @tolusha

### CODE REVIEW

File: plugins/plugin-java/che-plugin-java-ext-debugger-java-client/src/main/java/org/eclipse/che/ide/ext/java/jdi/client/debug/DebuggerPresenter.java [hunk 1]
```diff
@@ -148,13 +149,14 @@
     private       WorkspaceAgent                         workspaceAgent;
     private       FqnResolverFactory                     resolverFactory;
     private       EditorAgent                            editorAgent;
-    private       DebuggerVariable                       selectedVariable;
-    private       NotificationManager                    notificationManager;
+    private DebuggerVariable                       selectedVariable;
```
Comment (sleshchenko) [inline plugins/plugin-java/che-plugin-java-ext-debugger-java-client/src/main/java/org/eclipse/che/ide/ext/java/jdi/client/debug/DebuggerPresenter.java]: formatting

File: plugins/plugin-java/che-plugin-java-ext-debugger-java-server/src/main/java/org/eclipse/che/ide/ext/java/jdi/server/Debugger.java [hunk 2]
```diff
@@ -386,45 +390,47 @@ public void resume() throws DebuggerException {
...
-                                                   .withVariablePath(
-                                                           DtoFactory.getInstance().createDto(VariablePath.class)
-                                                                     .withPath(Collections.singletonList(var.getName()))
-                                                                    )
-                                                   .withPrimitive(var.isPrimitive()));
-        }
-        return dump;
+        synchronized (this) {
+            final JdiStackFrame currentFrame = getCurrentFrame();
+            StackFrameDump dump = DtoFactory.getInstance().createDto(StackFrameDump.class);
```
Comment (sleshchenko) [inline plugins/plugin-java/che-plugin-java-ext-debugger-java-server/src/main/java/org/eclipse/che/ide/ext/java/jdi/server/Debugger.java]: DtoFactory.getInstance().createDto(...) -> DtoFactory.newDto

File: plugins/plugin-java/che-plugin-java-ext-debugger-java-server/src/main/java/org/eclipse/che/ide/ext/java/jdi/server/Debugger.java [hunk 3]
```diff
@@ -115,6 +117,8 @@ public static Debugger getInstance(String name) {
     private ThreadReference thread;
     /** Current stack frame. Not <code>null</code> is thread suspended, e.g breakpoint reached. */
     private JdiStackFrame   stackFrame;
+    /**Lock for breakpoint steps*/
```
Comment (azatsarynnyy) [inline plugins/plugin-java/che-plugin-java-ext-debugger-java-server/src/main/java/org/eclipse/che/ide/ext/java/jdi/server/Debugger.java]: comment not formatted

**Comment by AndrienkoAleksandr** [comment]
@sleshchenko @skabashnyuk @vparfonov @tolusha

File: plugins/plugin-java/che-plugin-java-ext-debugger-java-server/src/main/java/org/eclipse/che/ide/ext/java/jdi/server/Debugger.java [hunk 4]
```diff
@@ -630,25 +644,19 @@ private boolean processBreakPointEvent(com.sun.jdi.event.BreakpointEvent event)
...
-                                                                               .withLocation(
-                                                                                       DtoFactory.getInstance().createDto(Location.class)
-                                                                                                 .withClassName(
-                                                                                                         location.declaringType().name())
-                                                                                                 .withLineNumber(location.lineNumber())
-                                                                                            )
-                                                                            )
-                                                             .withType(DebuggerEvent.BREAKPOINT);
+                breakPointEvent = (BreakPointEvent)newDto(BreakPointEvent.class)
+                        .withBreakPoint(
```
Comment (sleshchenko) [inline plugins/plugin-java/che-plugin-java-ext-debugger-java-server/src/main/java/org/eclipse/che/ide/ext/java/jdi/server/Debugger.java]: check formatting here

**Comment by sleshchenko** [comment]
other LGTM

**Comment by tolusha** [comment]
+1

**Comment by vparfonov** [comment]
+1

**Pull request merged**